src/Mvc/Mvc.Razor/src/HelperResult.cs

```diff
@@ -24,13 +24,21 @@ public HelperResult(Func<TextWriter, Task> asyncAction)
     {
         ArgumentNullException.ThrowIfNull(asyncAction);
 
-        _asyncAction = asyncAction;
+        // Hopefully people don't really do async work in the callback, but previously we were just blocking in WriteTo instead.
```

Comment:
You'd have to update the doc-comments on the constructor and the `WriteAction` property to reflect that this now runs inline.